### PR TITLE
[Driver][SYCL] Pass full set of sycl-post-link options to linker wrapper

### PR DESCRIPTION
**Pull request by mdtoguchi**
When using the new offloading model, pass all of the options associated with sycl-post-link to the clang-linker-wrapper.  These options will be used by the clang-linker-wrapper to setup the sycl-post-link calls during the device linking step.

This is more of a first step/general application here, as there are nuances in the options that should be used based on the device objects themselves.  The options passed are based on what is known by the driver specifically for that particular invocation.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -10594,39 +10587,33 @@ static std::string getSYCLPostLinkOptimizationLevel(const ArgList &Args) {
...
     else if (CodeSplitValue == "per_source")
-      addArgs(CmdArgs, TCArgs, {"-split=source"});
+      addArgs(PostLinkArgs, TCArgs, {"-split=source"});
     else if (CodeSplitValue == "auto")
-      addArgs(CmdArgs, TCArgs, {"-split=auto"});
+      addArgs(PostLinkArgs, TCArgs, {"-split=auto"});
     else { // Device code split is off
     }
-  } else if (T.getArchName() != "spir64_fpga") {
+  } else if (Triple.getArchName() != "spir64_fpga") {
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:10613]: I think we can try to move all 'Triple' related options to a separate function getTripleRelatedSYCLPostLinkOpts(....) and call it selectively for the old offload model.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -10635,62 +10622,85 @@ void SYCLPostLink::ConstructJob(Compilation &C, const JobAction &JA,
   // TODO: Try to extend this feature for non-Intel GPUs.
   if (!TCArgs.hasFlag(options::OPT_fno_sycl_remove_unused_external_funcs,
                       options::OPT_fsycl_remove_unused_external_funcs, false) &&
-      !T.isNVPTX() && !T.isAMDGPU() &&
-      !isSYCLNativeCPU(getToolChain()))
-    addArgs(CmdArgs, TCArgs, {"-emit-only-kernels-as-entry-points"});
+      !Triple.isNVPTX() && !Triple.isAMDGPU() && !isSYCLNativeCPU(TC))
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:10625]: We can use the triple checks selectively for old offloading model. We can move this to the getTripleRelatedSYCLPostLinkOpts function as well.

File: clang/test/Driver/sycl-intelfpga-aoco.cpp [hunk 3]
```diff
@@ -58,7 +58,7 @@
 // CHK-FPGA-AOCO: spirv-to-ir-wrapper{{.*}} "[[LIBLIST]]" "-o" "[[LIBLIST2:.+\.txt]]"
 // CHK-FPGA-AOCO: llvm-link{{.*}} "-o" "[[LINKEDBC:.+\.bc]]"
 // CHK-FPGA-AOCO: llvm-link{{.*}} "--only-needed" "[[LINKEDBC]]" "@[[LIBLIST2]]" "-o" "[[LINKEDBC2:.+\.bc]]"
-// CHK-FPGA-AOCO: sycl-post-link{{.*}} "-split-esimd"{{.*}} "-O2" "-spec-const=emulation" "-device-globals" "-o" "[[SPLTABLE:.+\.table]]" "[[LINKEDBC2]]"
```
Comment (asudarsa) [inline clang/test/Driver/sycl-intelfpga-aoco.cpp:61]: Hmm..I thought this will be a NFC. Why are these options getting removed? Thanks

Comment (mdtoguchi) [reply clang/test/Driver/sycl-intelfpga-aoco.cpp:61]: Just some cleanup - the options themselves in the context of this test are not important.  Reducing the options passed here allows for less potential changes down the line if the options passed to sycl-post-link change.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 1 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp]: New function created to move triple related items into.

**Review by hchilama — APPROVED** [review]
LGTM

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, this looks ready to merge - thanks!

**Pull request merged**